torchmultimodal/modules/layers/quantization.py

```diff
@@ -24,24 +24,60 @@ class Quantization(nn.Module):
     Vector quantization was introduced in Oord et al. 2017 (https://arxiv.org/pdf/1711.00937.pdf)
     to generate high-fidelity images, videos, and audio data.
 
+    The embedding weights are trained with exponential moving average updates as described
+    in original paper.
+
     Args:
         num_embeddings (int): the number of vectors in the embedding space
         embedding_dim (int): the dimensionality of the embedding vectors
 
     Inputs:
-        x (Tensor): Tensor containing a batch of encoder outputs.
+        z (Tensor): Tensor containing a batch of encoder outputs.
                     Expects dimensions to be batch x channel x n dims.
     """
 
-    def __init__(self, num_embeddings: int, embedding_dim: int):
+    def __init__(
+        self,
+        num_embeddings: int,
+        embedding_dim: int,
+        decay: float = 0.99,
+        epsilon: float = 1e-7,
+    ):
         super().__init__()
         self.embedding_dim = embedding_dim
         self.num_embeddings = num_embeddings
 
         self.embedding = nn.Embedding(self.num_embeddings, self.embedding_dim)
-        self.embedding.weight.data.uniform_(
-            -1 / self.num_embeddings, 1 / self.num_embeddings
-        )
+        # These are used in EMA update of embedding weights as m, N, and lambda, respectively, from Oord et al.
+        self._code_avg = self.embedding.weight.detach().clone()
+        self._code_usage = torch.zeros(self.num_embeddings)
+        self._decay = decay
+        # Used in Laplace smoothing of code usage
+        self._epsilon = epsilon
+
+        # Flag to track if we need to initialize embedding with encoder output
+        self._is_embedding_init = False
+
+    def _init_embedding_and_preprocess(self, z: Tensor) -> Tuple[Tensor, Size]:
+        # Embedding should be initialized with random output vectors from the encoder
+        # on the first forward pass for faster convergence, as in VideoGPT (Yan et al. 2021)
+        #
+        # This requires preprocessing the encoder output, so return this as well.
+
+        self._is_embedding_init = True
+
+        # Get random flattened encoder outputs
+        encoded_flat, permuted_shape = self._preprocess(z)
+        idx = torch.randperm(encoded_flat.shape[0])
+        encoded_flat_rand = encoded_flat[idx][: self.num_embeddings]
+
+        # Initialize embedding and intermediate values for EMA updates
+        with torch.no_grad():
```

Comment:
Those are not model parameters but more like "meta data" or "state" of the module. Check out `nn.Module.register_buffer`.

Reply:
if the model needed to be fine tuned, would these intermediate values need to be restored in `state_dict` after pre-training? I also noticed that in the MUGEN implementation they also include the embedding itself in register buffer, but to me it makes more sense as model parameters.

Reply:
actually, I see that buffers should be used for parameters that are not trained by the optimizer, but they are still restored in state_dict when the model is loaded (relevant forum [post](https://discuss.pytorch.org/t/what-is-the-difference-between-register-buffer-and-register-parameter-of-nn-module/32723)). In that case it would make sense for the embedding to be in the buffer